### PR TITLE
Isaiah/playwright tests

### PR DESCRIPTION
**Pull request by IsaiahA21**
I added playwright test coverage for HomePage, Nav bar, and Project page. 
Lmk if the tests are adequate.

### CODE REVIEW

File: playwright.config.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    /* Test against mobile viewports. */
+    // {
+    //   name: 'Mobile Chrome',
+    //   use: { ...devices['Pixel 5'] },
+    // },
+    // {
+    //   name: 'Mobile Safari',
+    //   use: { ...devices['iPhone 12'] },
+    // },
```
Comment (Anidion) [inline playwright.config.ts:59]: We should enable mobile viewports as well, but this can be a separate PR as the tests will need to be changed to use the hamburger menu at different sizes

**Review by Anidion — APPROVED** [review]
Great work! This is a really good start
If you're interested in learning about CI/CD that can be your next task, but up to you :)

File: tests/navigationBar/NavBar.test.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * Test Team button navigates to the team page and shows Founder
+ */
+test('Test Team button navigates to the team page and shows Founder', async ({ page }) => {
+  await page.goto('https://techstartucalgary.com/');
+  await page.locator('label').click();
+  await page.getByText('Team', { exact: true }).click();
+  await expect(page.getByRole('heading', { name: 'Our Team' })).toBeVisible();
+  await expect(page.locator('.FounderSectionstyles__FounderImg-sc-1pgvjm2-0')).toBeVisible();
```
[the remark this answers is not in the document]
Comment (brian-ngyn) [reply tests/navigationBar/NavBar.test.ts:23]: adding data-testid should be fine here imo

File: playwright.config.ts (hunk 1 above)
Comment (IsaiahA21) [reply playwright.config.ts:59]: Oh, that's very true.  
I think because of my screen size, when I started using record, it defaulted to the hamburger menu. So the tests in `NavBar.test.ts`, for example,  include: ` await page.locator('label').click();`

What Im going to do is get record to default away from the Hamburger menu and then modify the test cases. Then I'll create a new pr for mobile/hamburger menu

File: tests/navigationBar/NavBar.test.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+/**
+ * Test gallery button navigates to the Gallery page
+ */
+test('Test gallery button navigates to the Merch page', async function ({ page }) {
+  await page.goto('https://techstartucalgary.com/');
+  await page.locator('label').click();
+  await page.getByText('Gallery', { exact: true }).click();
+  await expect(page).toHaveURL('https://techstartucalgary.com/gallery');
+  await page.getByRole('heading', { name: 'Gallery' }).click();
+  await expect(page.locator('.sc-beySbM > img').first()).toBeVisible();
```
[the remark this answers is not in the document]
Comment (IsaiahA21) [reply tests/navigationBar/NavBar.test.ts:115]: done. 
In `PhotoGallery.tsx`, I added ``data-testid={`photo-gallery-image-${index}`}``

File: tests/navigationBar/NavBar.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (IsaiahA21) [reply tests/navigationBar/NavBar.test.ts:23]: > adding data-testid should be fine here imo

Thank you.
Inside `FounderSection.styles`  I've changed line 7 to ``export const FounderImg = styled.img.attrs({'data-testid': 'founder-img'})``

File: tests/navigationBar/NavBar.test.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+});
+
+/**
+ * Test Team button navigates to the team page and shows previous presidents
+ */
+test('Test Team button navigates to the team page and shows the 3 previous presidents', async ({ page }) => {
+  await page.goto('https://techstartucalgary.com/');
+  await page.locator('label').click();
+  await page.getByText('Team', { exact: true }).click();
+  await expect(page.getByRole('img', { name: '/static/media/Niyousha_Raeesinejad.7d226e07871c60102716.jpg' })).toBeVisible();
```
[the remark this answers is not in the document]
Comment (IsaiahA21) [reply tests/navigationBar/NavBar.test.ts:38]: Done.
In photogallery, include:
``const testDataId = `profile-image-${props.member.name.replace(/\s/g, "-").toLowerCase()}`;
`` to generate a unique id for each element.
Then did `data-testid={testDataId}`
Note: if you want to test it out, you need to switch the testcases to localhost

File: src/components/TeamFounder/FounderSection.styles.ts [hunk 5]
```diff
@@ -4,7 +4,7 @@ interface TeamStyledProps {
     mobileView: boolean;
 }
 
-export const FounderImg = styled.img`
+export const FounderImg = styled.img.attrs({'data-testid': 'founder-img'})`
```
Comment (brian-ngyn) [inline src/components/TeamFounder/FounderSection.styles.ts:7]: shouldn't we add data-testid to FounderSection.tsx?

File: src/components/TeamSection/Profile.tsx [hunk 6]
```diff
@@ -22,7 +22,7 @@ type ProfileProps = {
 const Profile = (props: ProfileProps) => {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   const preventDragHandler = (e: any) => e.preventDefault();
-
+  const testDataId = `profile-image-${props.member.name.replace(/\s/g, "-").toLowerCase()}`;
```
Comment (brian-ngyn) [inline src/components/TeamSection/Profile.tsx:25]: this value should be memoized

Comment (IsaiahA21) [reply src/components/TeamSection/Profile.tsx:25]: I've implemented the change but I'm not entirely sure why it's necessary. From my understanding, `useMemo` is useful during a component's lifecycle to avoid recalculating values on every re-render. It will only recalculate when the dependencies change.

However, in our case, we're mapping over `teamMembers`, so a Profile component is created for each member. Since `Profile.tsx` is already memoized using `React.memo()`, it only re-renders when its props change.

If the dependency for `useMemo` is `member.name`, it seems like, for each individual `Profile.tsx`, it doesn't add any significant performance improvement. Could you clarify why we need it in this scenario?

Comment (brian-ngyn) [reply src/components/TeamSection/Profile.tsx]: Memoizing a component helps prevent unnecessary re-renders when the component's props haven't changed. However, it does not prevent the recomputation of variables defined inside the component. Each time the parent component re-renders, the function body of the child component (including all variable definitions, hooks, etc.) will be re-executed, even if the component itself doesn't re-render.

This is because the memoization controls the re-rendering of the component's output (the JSX), but every time the parent re-renders, React still invokes the component function itself, which results in those internal variables being recomputed.

Comment (IsaiahA21) [reply src/components/TeamSection/Profile.tsx]: Thanks, i understand the difference between the 2 a lot clearer now

Comment (IsaiahA21) [reply src/components/TeamSection/Profile.tsx]: I’ve added you to the exclusive list of reviewers. If the exchanges are adequate could you pls approve it🥺

File: tests/homePage/HomePage.test.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+
+/**
+ * Test join team button
+ */
+test('join button should navigate to Apply page', async function ({ page }) {
+    await page.goto("https://techstartucalgary.com");
+
+    // find the join team button
+    await page.getByRole('link', { name: 'theTeam.join()' }).click();
```
Comment (Sukritib13) [inline tests/homePage/HomePage.test.ts:24]: This test is failing for firefox, I think its because firefox is more strict with text matching. For flexibility you can use regex, something like this - `await page.getByRole('link', { name: /theTeam\.join\(\)/ }).click();`

File: tests/navigationBar/NavBar.test.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  await page.getByText('About').click();
+  await expect(page).toHaveURL('https://techstartucalgary.com/');
+  await page.close();
+
+});
+
+/**
+ * Test Team button navigates to the team page and shows Founder
+ */
+test('Test Team button navigates to the team page and shows Founder', async ({ page }) => {
```
Comment (Sukritib13) [inline tests/navigationBar/NavBar.test.ts:17]: Could you attach a screenshot showing all the tests passing? A few of these tests are not passing on my end and I would like to crosscheck which ones before investigating more

File: content-updates.md [hunk 9]
```diff
@@ -23,6 +23,10 @@ This guide documents how and when to update different content on the Tech Start
     - [Applications](#applications)
   - [Resources Page](#resources-page)
     - [Guides](#guides)
+  - [Plawright Tests](#plawright-tests)
```
Comment (brian-ngyn) [inline content-updates.md:26]: There's a few typos here

File: content-updates.md [hunk 10]
```diff
@@ -220,3 +222,42 @@ Example:
...
+Here are 2 helpful links to get you started with Plawright
+[Plawright Getting Started ](https://playwright.dev/docs/intro) Or [Getting Started with Playwright and VS Code](https://www.youtube.com/watch?v=Xz6lhEzgI5I). 
+The second link is a video tutorial that will guide you through the process of setting up Plawright with VS Code, writing tests, and running them using the IDE.
+
+### How to run the tests using the terminal
+
+This will run all the tests in the tests folder
+
+```bash
+npx playwright test
```
Comment (brian-ngyn) [inline content-updates.md:239]: we should probably change the test script in package.json to run this command instead

File: tests/homePage/HomePage.test.ts (hunk 7 above)
Comment (IsaiahA21) [reply tests/homePage/HomePage.test.ts:24]: Thanks. I've made the change

File: tests/navigationBar/NavBar.test.ts (hunk 8 above)
Comment (IsaiahA21) [reply tests/navigationBar/NavBar.test.ts:17]: After making the change from your other comment here is my result. You're right some of the tests fails.
image
image
image

Comment (IsaiahA21) [reply tests/navigationBar/NavBar.test.ts:17]: Im trying to make it pass. Im not sure why its failing now

Comment (IsaiahA21) [reply tests/navigationBar/NavBar.test.ts:17]: I just remembered, you have to run the tests on localhost. I added a much of `data-testid` which don't exist on techstart.com

Comment (IsaiahA21) [reply tests/navigationBar/NavBar.test.ts:17]: I've pushed my branch - this is the latest error Im getting
image

Comment (IsaiahA21) [reply tests/navigationBar/NavBar.test.ts:17]: Fixed the (hopefully. Used a promise seems to work
image

File: tests/homePage/HomePage.test.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
+import { test, expect } from '@playwright/test';
+
+/**
+ * Test that logo shows up
+ */
+
+test('Should display Logo', async ({ page }) => {
+    await page.goto("https://techstartucalgary.com");
```
Comment (brian-ngyn) [inline tests/homePage/HomePage.test.ts:8]: There should be a helper function in the /tests folder to figure out what env we're on and navigate to the correct page based on dev or prd environment.

Consequently that raises the question on how we should integrate these tests within our CI.

File: content-updates.md (hunk 10 above)
Comment (IsaiahA21) [reply content-updates.md:239]: Can you explain what you mean pls.
Are you referring to changing line 48 in `package.json`?
image

File: tests/homePage/HomePage.test.ts (hunk 11 above)
Comment (IsaiahA21) [reply tests/homePage/HomePage.test.ts:8]: I created an env variable called `STAGING`. It's a toggle between 0 and 1. 1 for local env and 0 for prod.
Im running the test on `STAGING =1` And all the tests passed for Chrome, but on Firefox the same join button on the apply page is having an issue. My assumption is that it has to do with the animation on the button.

image
image

**Comment by brian-ngyn** [comment]
I'm seeing the following tests fail:

image

**Comment by IsaiahA21** [comment]
> I'm seeing the following tests fail:
> 
> image

Odd. were you in local or prod env?

**Comment by IsaiahA21** [comment]
image

Brian, when `STAGING=1`, these are the failures Im getting

File: playwright.config.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+  retries: process.env.CI ? 2 : 0,
+  /* Opt out of parallel tests on CI. */
+  workers: process.env.CI ? 1 : undefined,
+  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
+  reporter: 'html',
+  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
+  use: {
+    /* Base URL to use in actions like `await page.goto('/')`. */
+    // baseURL: 'http://127.0.0.1:3000',
+    baseURL: process.env.STAGING === '1' ? "http://localhost:3000" : "https://techstartucalgary.com",
```
[the remark this answers is not in the document]
Comment (IsaiahA21) [reply playwright.config.ts]: Thanks, Brian, U have added the GitHub workflow

[the remark this answers is not in the document]
Comment (IsaiahA21) [reply playwright.config.ts]: the report is generated and saved as an artifact

File: .github/workflows/playwright_tests_on_vercel.yml
Comment (Anidion) [inline .github/workflows/playwright_tests_on_vercel.yml:1]: Could you give this a nicer name? It shows up like this in GH workflows:

image

File: .github/workflows/playwright_tests_on_vercel.yml [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+      # Checkout the code from the repository
+      - uses: actions/checkout@v2
+
+      # Install project dependencies and Playwright (including browsers)
+      - name: Install dependencies
+        run: npm ci && npx playwright install --with-deps
+
+        # Run Playwright tests
+      - name: Run tests
+        run: npx playwright test --workers=4 --project=chromium
```
Comment (Anidion) [inline .github/workflows/playwright_tests_on_vercel.yml:19]: Not necessary now but we should also run it on Firefox and especially on mobile browsers. This can be a future task though

File: content-updates.md (hunk 9 above)
Comment (Anidion) [reply content-updates.md:26]: @IsaiahA21 let's address these before merging
```suggestion
  - [Playwright Tests](#playwright-tests)
```

File: content-updates.md [hunk 14]
```diff
@@ -23,6 +23,10 @@ This guide documents how and when to update different content on the Tech Start
     - [Applications](#applications)
   - [Resources Page](#resources-page)
     - [Guides](#guides)
+  - [Plawright Tests](#plawright-tests)
+    - [Installing Plawright](#installing-plawright)
```
Comment (Anidion) [inline content-updates.md:27]:
```suggestion
    - [Installing Playwright](#installing-playwright)
```

File: content-updates.md [hunk 15]
```diff
@@ -220,3 +222,42 @@ Example:
...
 />
 ```
+
+## Plawright Tests
+
+### Installing Plawright
+
+Here are 2 helpful links to get you started with Plawright
+[Plawright Getting Started ](https://playwright.dev/docs/intro) Or [Getting Started with Playwright and VS Code](https://www.youtube.com/watch?v=Xz6lhEzgI5I). 
+The second link is a video tutorial that will guide you through the process of setting up Plawright with VS Code, writing tests, and running them using the IDE.
```
Comment (Anidion) [inline content-updates.md:232]:
```suggestion
## Playwright Tests

### Installing Playwright

Here are 2 helpful links to get you started with Playwright
[Playwright Getting Started ](https://playwright.dev/docs/intro) Or [Getting Started with Playwright and VS Code](https://www.youtube.com/watch?v=Xz6lhEzgI5I). 
The second link is a video tutorial that will guide you through the process of setting up Playwright with VS Code, writing tests, and running them using the IDE.
```

File: global-setup.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+import dotenv from 'dotenv';
+import path from 'path';
+
+/**
+ * globalSetup configuration file to set something up once before running all tests
+ */
+
+async function globalSetupFunc() {
+    // import dotenv from 'dotenv';
+    // read from ".env.testEnv file"
```
Comment (Anidion) [inline global-setup.ts:11]: Let's keep dead code out
```suggestion
```

File: playwright.config.ts [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+    //   use: { ...devices['Desktop Chrome'], channel: 'chrome' },
+    // },
+  ],
+
+  /* Run your local dev server before starting the tests */
+  // webServer: {
+  //   command: 'npm run start',
+  //   url: 'http://127.0.0.1:3000',
+  //   reuseExistingServer: !process.env.CI,
+  // },
```
Comment (Anidion) [inline playwright.config.ts:85]:
```suggestion
```

File: src/components/TeamSection/Profile.tsx [hunk 18]
```diff
@@ -22,7 +22,11 @@ type ProfileProps = {
 const Profile = (props: ProfileProps) => {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   const preventDragHandler = (e: any) => e.preventDefault();
-
+  
+  // memoize the test data id so its not recalcaulated on every render but only when the member name changes
+  const testDataId = useMemo( () => {
+    return `profile-image-${props.member.name.replace(/\s/g, "-").toLowerCase()}`
+  },  [props.member.name]);
```
Comment (Anidion) [inline src/components/TeamSection/Profile.tsx:29]:
```suggestion
  const testDataId = useMemo(() => 
    `profile-image-${props.member.name.replace(/\s/g, "-").toLowerCase()}`,
  [props.member.name]);
```

File: src/pages/ProjectsPage.tsx [hunk 19]
```diff
@@ -108,13 +108,15 @@ const ProjectsPage = () => {
...
-          />
+          <div data-testid="apply_now_button">
+            <HoverButton
+              mode={ButtonMode.GRADIENT}
+              glowOnHover={true}
+              link="/apply"
+              text={"Apply Now"}
+              linkIsInternal={true}
+            />
+          </div>
```
Comment (Anidion) [inline src/pages/ProjectsPage.tsx:119]: I'd prefer passing the testid to the component but this is okay for now

**Review by Anidion — CHANGES_REQUESTED** [review]
Very minor changes then this will be good to go

**Review by Anidion — APPROVED** [review]
Looks good!
